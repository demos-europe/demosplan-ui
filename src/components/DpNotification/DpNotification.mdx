import { Canvas, Meta } from '@storybook/blocks'
import * as DpNotificationStories from './DpNotification.stories.tsx'

<Meta of={DpNotificationStories} />

# Notification

The Notification component is used to display brief messages or alerts to users. These messages typically provide feedback, convey information, or alert users about specific events or actions.

## Features

- Four notification types: info, warning, error, and confirm
- Auto-dismissal functionality (except for errors)
- Close button for manual dismissal
- Support for including links within notifications
- Consistent styling with appropriate icons
- Accessibility support

## Notification Types

### Info

Use for general information, updates, or non-critical messages.

<Canvas of={DpNotificationStories.Info} />

### Warning

Use to alert users about potential issues or actions that require caution.

<Canvas of={DpNotificationStories.Warning} />

### Error

Use to indicate errors or critical issues that need attention. These notifications do not auto-dismiss.

<Canvas of={DpNotificationStories.Error} />

### Confirm

Use to provide positive feedback or confirmation of successful actions.

<Canvas of={DpNotificationStories.Confirm} />

## Advanced Usage

### Notification with Link

This example demonstrates a notification with link functionality, which allows users to navigate to related resources.

<Canvas of={DpNotificationStories.WithLink} />