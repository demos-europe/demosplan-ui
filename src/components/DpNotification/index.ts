export { default } from './DpNotification.vue'
export * from './DpNotification.vue'
