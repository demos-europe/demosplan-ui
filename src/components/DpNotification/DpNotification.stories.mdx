import { Meta } from '@storybook/addon-docs'
import DpNotification from './'

<Meta
    title="Components/Notification"
    component={DpNotification}
/>

# Notification

The Notification component used to display brief messages or alerts to users.

These messages are typically used to provide feedback, convey information, or alert users about specific events or actions.

## Custom usage

```html
<dp-notification
    v-for="message in messages"
    :key="message.uid"
    :message="message"
    @dp-notify-remove="removeMessage()" />
```
