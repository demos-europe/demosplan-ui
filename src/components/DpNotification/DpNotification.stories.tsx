import type { Meta, StoryObj } from '@storybook/vue3-vite'
import DpNotification from './'

interface NotificationMessage {
  type: 'info' | 'warning' | 'error' | 'confirm' | 'dev'
  text: string
  linkUrl?: string
  linkText?: string
  uid?: string
}

interface IDpNotification {
  message: NotificationMessage
  hideTimer?: number
  'dp-notify-remove': (message: NotificationMessage) => void
}

const meta: Meta<typeof DpNotification> = {
  component: DpNotification,
  title: 'Components/Notification',
  render: (args) => ({
    components: { DpNotification },
    setup() {
      return { args }
    },
    template: `
      <div style="width: 400px; margin: 20px;">
        <dp-notification v-bind="args" @dp-notify-remove="args['dp-notify-remove']" />
      </div>
    `
  })
}

export default meta
type Story = StoryObj<IDpNotification>

export const Info: Story = {
  args: {
    message: {
      type: 'info',
      text: 'This is an information message'
    },
    hideTimer: 20000
  },
  argTypes: {
    'dp-notify-remove': { action: 'dp-notify-remove' }
  },
  parameters: {
    docs: {
      description: {
        story: 'Information notification with blue styling'
      }
    }
  }
}

export const Warning: Story = {
  args: {
    message: {
      type: 'warning',
      text: 'Warning! This action may have consequences'
    }
  },
  argTypes: {
    'dp-notify-remove': { action: 'dp-notify-remove' }
  },
  parameters: {
    docs: {
      description: {
        story: 'Warning notification with orange styling'
      }
    }
  }
}

export const ErrorState: Story = {
  name: 'Error',
  args: {
    message: {
      type: 'error',
      text: 'An error occurred while processing your request'
    }
  },
  argTypes: {
    'dp-notify-remove': { action: 'dp-notify-remove' }
  },
  parameters: {
    docs: {
      description: {
        story: 'Error notification with red styling. Error notifications do not auto-dismiss.'
      }
    }
  }
}

export const Confirm: Story = {
  args: {
    message: {
      type: 'confirm',
      text: 'Your changes have been successfully saved'
    }
  },
  argTypes: {
    'dp-notify-remove': { action: 'dp-notify-remove' }
  },
  parameters: {
    docs: {
      description: {
        story: 'Confirmation notification with green styling'
      }
    }
  }
}

export const WithLink: Story = {
  args: {
    message: {
      type: 'info',
      text: 'Documentation is available:',
      linkUrl: 'https://example.com/docs',
      linkText: 'View Documentation'
    },
    hideTimer: 20000
  },
  argTypes: {
    'dp-notify-remove': { action: 'dp-notify-remove' }
  },
  parameters: {
    docs: {
      description: {
        story: 'Notification with a clickable link'
      }
    }
  }
}
