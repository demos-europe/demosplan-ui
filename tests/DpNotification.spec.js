import DpNotification from '~/components/DpNotification'
import { mount } from '@vue/test-utils'

describe('DpNotification', () => {
  it('is named DpNotification', () => {
    expect(DpNotification.hasOwnProperty('name')).toBe(true)
    expect(DpNotification.name).toBe('DpNotification')
  })

  it('chooses the message class based on it\'s type', () => {
    const validCombinations = [
      {
        type: 'confirm',
        clazz: 'c-notify__message--confirm'
      },
      {
        type: 'info',
        clazz: 'c-notify__message--info'
      },
      {
        type: 'warning',
        clazz: 'c-notify__message--warning'
      }
    ]

    for (const val of validCombinations) {
      const wrapper = mount(DpNotification, {
        props: {
          message: {
            type: val.type
          }
        }
      })

      expect(wrapper.vm.messageClass).toBe(val.clazz)
    }
  })

  it('always renders with closemark', () => {
    let wrapper = mount(DpNotification, {
      props: {
        message: {
          type: 'error'
        }
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
    let closer = wrapper.find('i.c-notify__closer')
    expect(closer.element.tagName).toStrictEqual('I')

    wrapper = mount(DpNotification, {
      props: {
        message: {
          type: 'confirm'
        }
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
    closer = wrapper.find('i.c-notify__closer')
    expect(closer.element.tagName).toStrictEqual('I')
  })

  it('renders the message text', () => {
    const wrapper = mount(DpNotification, {
      props: {
        message: {
          type: 'confirm',
          text: 'MessageText'
        }
      }
    })

    expect(wrapper.find('.flow-root > .u-ml').text()).toBe('MessageText')
  })

  it('renders a link if link attributes are given', () => {
    const wrapper = mount(DpNotification, {
      props: {
        message: {
          type: 'confirm',
          text: 'MessageText',
          linkUrl: 'about:blank',
          linkText: 'LinkText'
        }
      }
    })

    expect(wrapper.find('.flow-root > .u-ml').text()).toBe('MessageText LinkText')
  })

  /*
   * For some reason this Test takes over 6 seconds to pass
   */
  /*
   * it('emits dp-notify-remove with its message as payload once clicked', (done) => {
   *   const message = {
   *     type: 'confirm',
   *     text: 'MessageText'
   *   }
   * 
   *   const wrapper = mount(DpNotification, {
   *     props: {
   *       message: message,
   *       hideTimer: 2
   *     }
   *   })
   * 
   *   wrapper.find('.c-notify__closer').trigger('click')
   * 
   *   expect(wrapper.emitted()).toHaveProperty('dp-notify-remove')
   *   expect(wrapper.emitted()['dp-notify-remove'][0]).toEqual([message])
   * })
   */

  /*
   * For some reason this Test takes over 6 seconds to pass
   */
  /*
   * it('emits dp-notify-remove after hide timeout idle', (done) => {
   *   const message = { type: 'confirm', text: 'MessageText' }
   * 
   *   const wrapper = mount(DpNotification, {
   *     props: {
   *       message: message,
   *       hideTimer: 25 // Make this timeout reasonably short to keep the test time low
   *     }
   *   })
   * 
   *   console.log('test notification')
   * 
   *   jest.setTimeout(() => {
   *     console.log('test notification timeout ', wrapper.emitted()['dp-notify-remove'][0])
   *     expect(wrapper.emitted()).toHaveProperty('dp-notify-remove')
   *     expect(wrapper.emitted()['dp-notify-remove'][0]).toEqual([message])
   * 
   *   }, 27)
   * }, 30000)
   */
})
