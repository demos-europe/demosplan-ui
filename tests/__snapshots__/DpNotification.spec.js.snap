// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DpNotification always renders with closemark 1`] = `
"<div class="c-notify__message c-notify__message--error"><i class="c-notify__icon c-notify__closer fa fa-times-circle cursor-pointer" aria-hidden="true"></i>
  <div class="flow-root"><i class="c-notify__icon fa u-mt-0_125 u-mr-0_25 float-left fa-exclamation-circle"></i>
    <div class="u-ml">
      <!--v-if-->
    </div>
  </div>
</div>"
`;

exports[`DpNotification always renders with closemark 2`] = `
"<div class="c-notify__message c-notify__message--confirm"><i class="c-notify__icon c-notify__closer fa fa-times-circle cursor-pointer" aria-hidden="true"></i>
  <div class="flow-root"><i class="c-notify__icon fa u-mt-0_125 u-mr-0_25 float-left fa-check-circle"></i>
    <div class="u-ml">
      <!--v-if-->
    </div>
  </div>
</div>"
`;
